### PR TITLE
feat: Immiscible multiphase flow

### PR DESCRIPTION
**Pull request by rpiazza87**
Implementation of a multiphase immiscible solver for compressible fluids with viscous, gravity and capillary forces:

- Included new `ImmiscibleMultiphaseFlow` solver and associated kernels (currently limited to two phases)
- Standard and total mass formulations
- Cell-centered Dirichlet and flux boundary conditions
- Included new `TwoPhaseFluid` model for tabular data input
- Direct and iterative linear solver compatibility
- GPU compatibility
- Set of input files with validation cases
- Unit tests and documentation provided

### CODE REVIEW

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreMGR.hpp [hunk 1]
```diff
@@ -213,6 +213,7 @@ class MGRStrategyBase
 #if GEOS_USE_HYPRE_DEVICE == GEOS_USE_HYPRE_CUDA || GEOS_USE_HYPRE_DEVICE == GEOS_USE_HYPRE_HIP
     GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetAggInterpType( solver.ptr, hypre::getAMGAggressiveInterpolationType( LinearSolverParameters::AMG::AggInterpType::modifiedExtendedE ) ) );
     GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetCoarsenType( solver.ptr, hypre::getAMGCoarseningType( LinearSolverParameters::AMG::CoarseningType::PMIS ) ) );
+    GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetAggNumLevels( solver.ptr, 0 ) );
```
Comment (CusiniM) [inline src/coreComponents/linearAlgebra/interfaces/hypre/HypreMGR.hpp]: can you remove this and create a PR only with this change? @victorapm we should test this. Unless you know why this would impact GPU performance so much.

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  m_tableFiles.empty() ? readInputDataFromTableFunctions() : readInputDataFromFileTableFunctions();
+
+  checkTableConsistency();
+  /* TODO
+     // call to correctly set member array tertiary sizes on the 'main' material object
+     resizeFields( 0, 0 );
+
+     // set labels on array wrappers for plottable fields
+     setLabels();
+   */
```
Comment (CusiniM) [inline src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp]: remove?

**Review by paveltomin — APPROVED** [review]

**Review by victorapm — APPROVED** [review]
The new MGR strategy looks good, thank you!

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreMGR.hpp (hunk 1 above)
Comment (victorapm) [reply src/coreComponents/linearAlgebra/interfaces/hypre/HypreMGR.hpp]: Agreed on testing this. It seems aggressive coarsening did not behave well for some of the problems tested in this PR, while it works well for other problems I tested. For better robustness, we can leave it turned off by default as proposed in this PR, but ultimately we should add the option of controlling this via input file

File: src/coreComponents/constitutive/unitTests/testTwoPhaseFluid.cpp [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+#include "constitutive/fluid/twophasefluid/TwoPhaseFluid.hpp"
+#include "constitutive/fluid/twophasefluid/TwoPhaseFluidFields.hpp"
+
+// Only for fill
+#include "unitTests/constitutiveTests/MultiFluidTest.hpp"
```
Comment (dkachuma) [inline src/coreComponents/constitutive/unitTests/testTwoPhaseFluid.cpp]: I think this might be introducing a dependency on `unitTests/constitutiveTests`. Maybe this test should move to `unitTests/constitutiveTests`.

File: src/coreComponents/constitutive/docs/TwoPhaseFluid.rst [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+Overview
+=========================
+
+This model represents a two-phase fluid with pressure-dependent density and viscosity.
+
+For each phase, both density and viscosity are described as tabulated data, either in the form of ``TableFunction`` or text files.
+
+In the case of text files, one file is expected per phase and should consist of three columns: pressure, density and viscosity.
+
+Note that currently, there is no temperature dependence in the model.
```
Comment (dkachuma) [inline src/coreComponents/constitutive/docs/TwoPhaseFluid.rst]: Am I correct in thinking that this can only be used with the immiscible solver? Can we add a note here about that.

**Review by dkachuma — APPROVED** [review]

**Comment by paveltomin** [comment]
@CusiniM, @MelReyCG, @corbett5, @rrsettgast, @wrtobin please review

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.hpp
Comment (tjb-ltk) [inline src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.hpp:1]: What is the motivation for splitting the 2d input tables (density and visc).  if the prop lookup is always at the same pressure then it seems like the index calculation is done twice, or more if considering derivatives..

File: src/coreComponents/constitutive/docs/TwoPhaseFluid.rst (hunk 4 above)
Comment (tjb-ltk) [reply src/coreComponents/constitutive/docs/TwoPhaseFluid.rst]: Along similar lines ... @dkachuma  should the fluid model be called TwoPhaseImmiscible ??

**Review by castelletto1 — APPROVED** [review]

File: inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_GravitySegregation_1d/initialSaturation1.txt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+0
+0
+0
+0
+0
```
Comment (CusiniM) [inline inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_GravitySegregation_1d/initialSaturation1.txt]: seems like this can be done with only 4 entries?

File: inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permx.geos
Comment (CusiniM) [inline inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permx.geos:1]: this file should not live in the geos repo but in GEOSDATA

File: inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permy.geos
Comment (CusiniM) [inline inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permy.geos:1]: same

File: inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permz.geos
Comment (CusiniM) [inline inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permz.geos:1]: same

File: inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/xlin.geos
Comment (CusiniM) [inline inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/xlin.geos:1]: same applies to these files.

File: src/coreComponents/constitutive/docs/TwoPhaseFluid.rst (hunk 4 above)
Comment (CusiniM) [reply src/coreComponents/constitutive/docs/TwoPhaseFluid.rst]: yes, IMO

File: inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permx.geos
Comment (castelletto1) [reply inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permx.geos:1]: Agreed, let's add a folder `Layer84` in https://github.com/GEOS-DEV/GEOSDATA/tree/develop/DataSets/SPE10 and move such files there.

File: src/coreComponents/constitutive/unitTests/testTwoPhaseFluid.cpp (hunk 3 above)
Comment (CusiniM) [reply src/coreComponents/constitutive/unitTests/testTwoPhaseFluid.cpp]: This include is an issue.

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 6]
```diff
@@ -0,0 +1,1431 @@
...
+                                                                             stencilWrapper,
+                                                                             dt,
+                                                                             localMatrix.toViewConstSizes(),
+                                                                             localRhs.toView() );
+    } );
+  } );
+}
+
+// Ryan: Looks like this will need to be overwritten as well...
+// I have left the CompositionalMultiphaseFVM implementation for reference
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp]: clean up some of this comments please.

**Review by CusiniM — APPROVED** [review]

**Review by tjb-ltk — APPROVED** [review]

**Review by jhuang2601 — APPROVED** [review]

File: src/coreComponents/common/DataLayouts.hpp [hunk 7]
```diff
@@ -21,6 +21,7 @@
 #define GEOS_COMMON_DATALAYOUTS_HPP_
 
 #include "common/GeosxConfig.hpp"
+#include "common/DataTypes.hpp"
```
Comment (MelReyCG) [inline src/coreComponents/common/DataLayouts.hpp]: Why? (it don't seem to be used here, and `DataTypes.hpp` includes `DataLayouts.hpp`, so maybe directly include `DataTypes.hpp` if needed)

File: src/coreComponents/common/DataLayouts.hpp [hunk 8]
```diff
@@ -205,6 +206,44 @@ static constexpr int STIFFNESS_USD = LvArray::typeManipulation::getStrideOneDime
...
+/// indices of pressure and saturation derivatives
+struct DerivativeOffset
+{
+  /// index of derivative wrt pressure
+  static int constexpr dP = 0;
+  /// index of first derivative wrt compositions
+  static int constexpr dS = 1;
+};
+
+} // namespace immiscibleFlow
```
Comment (MelReyCG) [inline src/coreComponents/common/DataLayouts.hpp:244]: A suggestion, why not locating those declarations "less globally"?
Here's what I mean:
1. the `constitutive/fluid/twophaseimmisciblefluid` could become `constitutive/fluid/immiscibleFluid`,
2. We would have a `constitutive/fluid/immiscibleFluid/Layout.hpp` containing this portion of code,
3. `ImmiscibleMultiphaseKernels.hpp` & `ImmiscibleMUltiphaseFlowFiels.hpp` would include this `Layout.hpp`.

No modification would be done at the very base of the codebase. IIUC, these memory layouts details are only locally used.

**Review by MelReyCG — COMMENTED** [review]
A few comments on the last file to approve.

File: src/coreComponents/constitutive/unitTests/testTwoPhaseFluid.cpp (hunk 3 above)
Comment (bd713) [reply src/coreComponents/constitutive/unitTests/testTwoPhaseFluid.cpp]: Problem solved by following @dkachuma 's suggestion.

File: src/coreComponents/constitutive/docs/TwoPhaseFluid.rst (hunk 4 above)
Comment (bd713) [reply src/coreComponents/constitutive/docs/TwoPhaseFluid.rst]: Agreed! The naming was made more explicit by following @tjb-ltk's suggestion.

File: inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permx.geos
Comment (bd713) [reply inputFiles/immiscibleMultiphaseFlow/immiscibleTwoPhase_SPE10_layer84/layer84_permx.geos:1]: As discussed and agreed during the April 3rd GEOS Infrastructure meeting, this won't be addressed in this PR. 
This should be part of a wider effort to restructure the tests for **ALL** solvers: move the data to the appropriate repository, promote HDF5 binary format as opposed to ASCII files, etc.

File: src/coreComponents/common/DataLayouts.hpp (hunk 8 above)
Comment (paveltomin) [reply src/coreComponents/common/DataLayouts.hpp:244]: it seems that this PR just follows current way how everything is done
i like your idea but it is probably out of scope here?

Comment (castelletto1) [reply src/coreComponents/common/DataLayouts.hpp:244]: I agree with @paveltomin. We should have a dedicated PR to implement such change, and not only for the new solvers added here.

Comment (castelletto1) [reply src/coreComponents/common/DataLayouts.hpp:244]: @MelReyCG : could you please approve so that we can move this PR to the Merge Queue?

Comment (rrsettgast) [reply src/coreComponents/common/DataLayouts.hpp:244]: we certainly can have a localized approach to things. perhaps the layouts should be in defined in the solver.

**Review by rrsettgast — APPROVED** [review]
